### PR TITLE
#12 | update readme with the section supported configuration

### PR DESCRIPTION
**Pull request by hoavt**
connect #12 

@hieptranquoc Please review this pr.

### CODE REVIEW

**Comment by hieptranquoc** [comment]
@hoavt It should be detailed, rather than listed. Specifies the function of each major configure.

File: README.md [hunk 1]
```diff
@@ -67,3 +67,57 @@ teracy-dev:
...
+
+### Examples
+
+This section gives you some examples about the supported configuration in teracy-dev-core, so you should know how
+to override the default configuration.
+
+You should add your new configuration into the `workspace/teracy-dev-entry/config_override.yaml`.
+
+*vagrant*
+
```
Comment (hieptranquoc) [inline README.md:99]: Specify what this config does and what it contains

File: README.md [hunk 2]
```diff
@@ -67,3 +67,57 @@ teracy-dev:
...
+Please see an example at `workspace/teracy-dev-essential/blob/develop/config.yaml`.
+
+*default*
+
```
Comment (hieptranquoc) [inline README.md:103]: need example config of `default`

File: README.md [hunk 3]
```diff
@@ -67,3 +67,57 @@ teracy-dev:
...
+This `default` configuration is set for all nodes to use. It consits of 4 configuration: `vm`, `providers`, `provisioners`,
+and `plugins`.
+ 
+- `providers`: see more at https://www.vagrantup.com/docs/providers/.
+- `provisioners`: by default, we have 2 provisions : `host`, and `guest`. You can see more at https://www.vagrantup.com/docs/provisioning/.
+- `plugins`: see more at https://www.vagrantup.com/docs/plugins/.
+
+
+*node*
```
Comment (hieptranquoc) [inline README.md]: Specify what this config does and what it contains.

**Comment by hieptranquoc** [comment]
@hoavt we can create `example` directory, and put `config.example.yaml` in it.

**Review by hieptranquoc — CHANGES_REQUESTED** [review]

**Comment by hoavt** [comment]
> @hoavt It should be detailed, rather than listed. Specifies the function of each major configure.

Their function are mentioned in the configuration file. Why do we need to rewrite them anymore? When we have any changes in the config file, we don;t need to change the README file anymore.

File: README.md (hunk 1 above)
Comment (hoavt) [reply README.md:99]: It is mentioned in the config.yaml, why do we need to mention it here anymore?

File: README.md (hunk 2 above)
Comment (hoavt) [reply README.md:103]: @hieptranquoc Please help me to give the example for it :)

File: README.md (hunk 3 above)
Comment (hoavt) [reply README.md]: It is mentioned in the config. yaml file. We don't  need to mention it here. @hoatle please confirm.

**Comment by hoatle** [comment]
it's expected that users should know how to use by referring docs, not the source code. Docs should explain in more details for what and why, which param is required, which is optional, what is the default value, etc.

**Comment by hoatle** [comment]
examples are expected for different config types: variables, plugins, default and nodes, etc.

**Comment by hoavt** [comment]
@hieptranquoc Please review, and add more examples for "default" and "variables". Moreover, which params need to be explained more, please help me to explain them. which param is required, which is optional...

**Review by hieptranquoc — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
         branch: develop
       require_version: ">= 0.4.0-SNAPSHOT"
 ```
+
+## Supported configuration
+These are the supported configuration in teracy-dev-core: `variables`, `vagrant`, `default`, `node`.
+
+
+### variables
+`variables` which is the environment variables must be defined to use `teracy-dev`. All variables
```
Comment (hoatle) [inline README.md]: `variables` is used to define dynamic config values. You can define variable keys with environment variables.

File: README.md [hunk 5]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+# and "%{key}" can be used for settings values if available
+variables:
+  node_name_prefix: ${NODE_NAME_PREFIX:-node}
+  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}
+  node_domain_affix: ${NODE_DOMAIN_AFFIX:-local}
+  # some_value: ${MY_VAR:-value}-affix
+```
+In which:
+
+- `node_name_prefix` param: its value is `NODE_NAME_PREFIX`. If `NODE_NAME_PREFIX` is not defined, the value
```
Comment (hoatle) [inline README.md]: its value will be the value of the env var `NODE_NAME_PREFIX` is defined, otherwise, ...

File: README.md [hunk 6]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+should be the default `node`.
+- `node_hostname_prefix` param: its value is `NODE_HOSTNAME_PREFIX`. If `NODE_HOSTNAME_PREFIX` is not defined, the value should be the default `node`.
+- `node_domain_affix` param: its value is `NODE_DOMAIN_AFFIX`. If `NODE_DOMAIN_AFFIX` is not defined, the value should be the default `local`.
+
+
+### vagrant
```
Comment (hoatle) [inline README.md]: this vagrant config belongs to https://github.com/teracyhq/dev (teracy-dev-kernel, not teracy-dev-core), move this.

File: README.md [hunk 7]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+    #   config_key: gatling
+    #   options:
+    #     latency: 0.5
+    #     time_format: "%H:%M:%S"
+    #     rsync_on_startup: true
+```
+
+### default
+
+The `default` settings which is used for all nodes and should be defined in the `default` block. It
```
Comment (hoatle) [inline README.md]: which will be overridden by each `nodes`, so any config which is shared among nodes, we should use `default`

File: README.md [hunk 8]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+consits of 4 configuration: `vm`, `providers`, `provisioners`, and `plugins`.
+
+- `vm`: Virtual machine setting 
+- `providers`: see more at https://www.vagrantup.com/docs/providers/.
+- `provisioners`: by default, we have 2 provisions : `host`, and `guest`. You can see more at
```
Comment (hoatle) [inline README.md]: incorrect info

File: README.md [hunk 9]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+      memory: 1024
+      description: "%{node_name_prefix} #{Time.now.getutc.to_i}"
+
+  provisioners: []
+
+  # default node config level
+  plugins: []
+```
+
+### node
```
Comment (hoatle) [inline README.md]: nodes

File: README.md [hunk 10]
```diff
@@ -67,3 +67,120 @@ teracy-dev:
...
+The `node` configure which is used to install each node and should be defined in the `nodes` block.
```
Comment (hoatle) [inline README.md]: is used to define config for each VMs

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc @hoavt check my comment, make sure to check and verify more carefully.

**Comment by hoavt** [comment]
@hieptranquoc Please review the new commit.

**Review by hieptranquoc — DISMISSED** [review]

File: README.md [hunk 11]
```diff
@@ -67,3 +67,92 @@ teracy-dev:
...
+variables:
+  node_name_prefix: ${NODE_NAME_PREFIX:-node}
+  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}
+  node_domain_affix: ${NODE_DOMAIN_AFFIX:-local}
+  # some_value: ${MY_VAR:-value}-affix
+```
+In which:
+
+- `node_name_prefix` param: its value will be the value of the env var `NODE_NAME_PREFIX` which is defined, otherwise, the value should be the default `node`.
+- `node_hostname_prefix` param: its value is `NODE_HOSTNAME_PREFIX`. If `NODE_HOSTNAME_PREFIX` is not defined, the value should be the default `node`.
```
Comment (hoatle) [inline README.md]: wrong info, see the explanation of `node_name_prefix`

File: README.md [hunk 12]
```diff
@@ -67,3 +67,92 @@ teracy-dev:
...
+- `node_domain_affix` param: its value is `NODE_DOMAIN_AFFIX`. If `NODE_DOMAIN_AFFIX` is not defined, the value should be the default `local`.
```
Comment (hoatle) [inline README.md]: wrong info, see the explanation of `node_name_prefix`

File: README.md [hunk 13]
```diff
@@ -67,3 +67,92 @@ teracy-dev:
...
+    vm:
+      hostname: "%{node_hostname_prefix}-01.%{node_domain_affix}"
+```
+
+### Examples
+
+This section gives you some examples about the supported configuration in teracy-dev-core, so you should know how
+to override the default configuration.
+
+Please see an example at `examples/config_example.yaml`.
```
Comment (hoatle) [inline README.md]: explain how to run the examples with realistic use cases that users can run the code

**Review by hoatle — CHANGES_REQUESTED** [review]
@hoavt @hieptranquoc pls check my comments

**Comment by hoavt** [comment]
@hieptranquoc please review.

**Review by hieptranquoc — APPROVED** [review]

**Comment by hoatle** [comment]
need to split the examples into the following example configs:
- examples/variables.yaml
- examples/plugins.yaml
- examples/vm-vb-providers.yaml
- examples/file-provisioner.yaml
- examples/shell-provisioner.yaml
- examples/ansible-local-provisioner.yaml
- examples/chef-solo-provisioner.yaml
- examples/multi-nodes.yaml

how to run and expected outcome and behavior should be explained.

**Pull request closed without merging**